src/common/nodes/CCIMSNode.ts

```diff
@@ -0,0 +1,143 @@
+import { NodeType } from "./NodeType";
+import { DatabaseManager } from "../database/DatabaseManager";
+import { DatabaseCommand } from "../database/DatabaseCommand";
+import { Saveable } from "./Saveable";
+import { NodeTableSpecification, RowSpecification } from "./NodeTableSpecification";
+import { DeleteNodeCommand } from "../database/commands/DeleteNodeCommand";
+import { AddNodeCommand } from "../database/commands/save/AddNodeCommands";
+import { UpdateNodeCommand } from "../database/commands/save/UpdateNodeCommand";
+
+export const CCIMSNodeTableSpecification: NodeTableSpecification<CCIMSNode>
+    = new NodeTableSpecification<CCIMSNode>("node", undefined, RowSpecification.fromProperty("id", "id"));
+
+/**
+ * Base class for all datatypes with an id, which are accessable via the api
+ */
+export abstract class CCIMSNode<T extends CCIMSNode = any> implements Saveable {
+    private _id: string;
+    private _type: NodeType;
+    private _isNew: boolean = false;
+    private _isChanged: boolean = false;
+    private _isDeleted: boolean = false;
+    protected databaseManager: DatabaseManager;
+    private _saveables: Saveable[] = [];
+    protected _tableSpecification: NodeTableSpecification<T>;
+
+    protected constructor(type: NodeType, databaseManager: DatabaseManager, tableSpecification: NodeTableSpecification<T>, id: string) {
+        this._id = id;
+        this._type = type;
+        this.databaseManager = databaseManager;
+        this._tableSpecification = tableSpecification;
+    }
+
+    /**
+     * gets the unique id, which is used for the database
+     * this is neither null nor empty
+     */
+    public get id(): string {
+        return this._id;
+    }
+
+    /**
+     * gets the type of this node
+     * this is also used for the graphql resolvers
+     */
+    public get type(): NodeType {
+        return this._type;
+    }
+
+    /**
+     * returns true, if this node was modified or newly created
+     */
+    public isChanged(): boolean {
+        return this._isChanged;
+    }
+
+    /**
+     * returns true if this node was newly created
+     */
+    public isNew(): boolean {
+        return this._isNew;
+    }
+
+    /**
+     * returnes true if this node is deleted
+     * this does not indicate that it was newly deleted
+     */
+    public isDeleted(): boolean {
+        return this._isDeleted;
+    }
+
+    /**
+     * marks this noda as new, and therefore also as changed
+     */
+    protected markNew(): void {
+        this._isNew = true;
+        this._isChanged = true;
+    }
+
+    /**
+     * marks this node as changed, aka isChanged will return true
+     */
+    public markChanged(): void {
```

Comment:
I cannot make this a property, because it is not allowed to set the value false
it is only possible to mark it as changed, but is is not possible to mark it as unchanged

Reply:
Ok